seps/sep-0002.md

```diff
@@ -0,0 +1,100 @@
+# Support multiple build-systems per package using a directive
+
+## The problem
+
+There are multiple packages that are either changing their build-system during the evolution of the project, or using different build-systems for different platforms. Spack, at the moment, provides no support to model this use case. As a result we wrote some [documentation](https://github.com/spack/spack/pull/25174) to enumerate various workarounds that people adopted - each with its own drawbacks. What we would like to have in the long term is proper support for packages that can be built using multiple build-systems.
+
+## Proposed changes
+
+What we propose here to add support for packages using multiple build-systems is to __add another directive__ to the package DSL. If we take as an example `hdf5` this would look like:
+```python
+class Hdf5(Package):
+    build_system('cmake', when='@X.Y:')
+    build_system('autotools')
+```
+Each build-system can be subject to a constraint. In the example above, for instance, `cmake` can be used only if the version of `hdf5` is greater or equal than `X.Y` (while `autotools` is always available). 
+
+We can also assume that there is an __implicit preference order__ based on on the order of declaration in the class. In the example above this means that, if no other requests are made by the user, `cmake` is preferred to `autotools` when available.
+
+Build systems can be specified in specs using a key-value pair:
+```
+hdf5 build_system=cmake
+```
+__which requires the keyword `build_system` to be reserved by Spack__ (similarly to what is done for `dev_path`). It is interesting to note that modeling build-systems this way make them similar to signle-valued variants with the only difference that the set of allowed values is "dynamic" and subject to constraints.
```

Comment:
```suggestion
__which requires the keyword `build_system` to be reserved by Spack__ (similarly to what is done for `dev_path`). It is interesting to note that modeling build-systems this way make them similar to single-valued variants with the only difference that the set of allowed values is "dynamic" and subject to constraints.
```
A **lot** of users have been asking for this for swapping between Ninja and Makefile in CMakePackage. Would you consider Ninja vs. Makefile to be a build system? We could do `build_system=cmake_ninja` or `build_system=cmake_makefile`. Or it could be a different reserved keyword.

Reply:
I have to think a bit about this. A solution like `build_system=cmake_ninja` does not require any change in the proposed semantics, but we are effectively splitting the `cmake` build-system into many build-systems based on the generator we use. 

Using `hdf5 build_system=cmake generator=ninja`seems more appealing, but we need the ability to attach specific variants based on the build-system being used. That may require the "conditional" variant PR from @becker33:
```python
variant('generator', values=('makefile', 'ninja'), default='makefile', when'build_system=cmake')
```